ext/libclementine-tagreader/tagreader.cpp

```diff
@@ -130,6 +131,52 @@ const char* kASF_OriginalDate_ID = "WM/OriginalReleaseTime";
 const char* kASF_OriginalYear_ID = "WM/OriginalReleaseYear";
 }  // namespace
 
+// Helper for GuessArtistAndTitle()
+static void removeExtension(std::string& s) {
+  if (s.empty()) return;
+  const int i = s.find_last_of('.');
+  if (i == std::string::npos) return;
+  s.assign(s.substr(0, i));
+}
+
+static void changeUnderscores(std::string& s) {
+  std::replace(s.begin(), s.end(), '_', ' ');
+}
+
+static std::string& ltrim(std::string& str, const std::string& chars = "\t\n\v\f\r ") {
+  str.erase(0, str.find_first_not_of(chars));
+  return str;
+}
+
+static std::string& rtrim(std::string& str, const std::string& chars = "\t\n\v\f\r ") {
+  str.erase(str.find_last_not_of(chars) + 1);
+  return str;
+}
+
+static std::string& trim(std::string& str, const std::string& chars = "\t\n\v\f\r ") {
+  return ltrim(rtrim(str, chars), chars);
+}
+
+void TagReader::GuessArtistAndTitle(pb::tagreader::SongMetadata *song) const {
+  std::string artist = song->artist();
+  std::string title = song->title();
+  const std::string bn = song->basefilename();
+  if (!artist.empty() || !title.empty()) return;
+  if (bn.empty()) return;
+  int i = bn.find(" - ");
+  if (i == std::string::npos) i = bn.find("_-_");
+  if (i == std::string::npos) return;
```

Comment:
Have you thought about using std::regex? You could match both separators and remove the extension in a single expression.

Reply:
ok will do